scripts/linter/valid-recipes.js

```diff
@@ -0,0 +1,40 @@
+const visit = require("unist-util-visit");
+
+const ruleId = "stumptown-linter:frontmatter-valid-recipe";
+
+function hasRecipeInYaml(node) {
+    return node.data && node.data.yaml && node.data.yaml.recipe;
+}
+
+/**
+ * If a tree contains a YAML (frontmatter) node and it contains a recipe, then confirm that it's a recognized recipe.
+ *
+ * As a convenience, if the recipe is valid, then add the recipe object itself to the root of the tree as `node.data.recipe`.
+ */
+function attacher(options) {
+    const recipes = options.recipes;
+
+    return function transformer(tree, file) {
+        visit(tree, "yaml", node => {
+            const validRecipes = Object.keys(recipes);
+            if (hasRecipeInYaml(node)) {
+                const recipe = node.data.yaml.recipe;
+
+                if (validRecipes.includes(recipe)) {
+                    tree.data = tree.data || {};
+                    tree.data.recipe = recipes[recipe];
+                    tree.data.recipeName = recipe;
+                } else {
+                    const m = file.message(
+                        `"${recipe}" recipe was not found in /recipes`,
```

Comment:
This errors for guide pages and landing pages, that have recipes named in front matter (`guide` and `landing-page`), but don't have YAML files, because they're not obviously amenable to the recipe.yaml model (because they don't have the same kind of structure as a reference page, which we can represent as a sequence of named sections).

Guide pages are just totally unstructured blobs of Markdown, with special bits (e.g. live samples) embedded in them at arbitrary places. Landing pages are a prose overview followed by a list of lists of links. So they're both treated as special cases, where the recipe name just tells the builder what to expect, without an actual recipe to consult.

I can think of about three options here:

* continue to live with this weirdness
* give guide pages and landing pages actual recipes, and use them. Now I think about this, we could perhaps write a useful recipe for landing pages. Still I think the guide page recipe would just be "some stuff". Another related problem is that at the moment, for reference pages, the recipe determines the sidebar (because all HTML element pages want the same sidebar), but guide pages and landing pages are generic across domains, so a "guide" recipe can't determine the sidebar.
* add some extra datum like "type", which could be ["guide"|"landing-page"|"reference"]. If type is "reference", expect "recipe" to exist.

FWIW I do expect that guide pages and landing pages would be the only pages that are not recipe-driven.

Reply:
Can we have recipe files that don't specify a `body`? Or even essentially emtpy recipe files that say, `# read whatever.js to see how this recipe is handled`?

It would be nice to have a single place to see what the valid recipes are, even if some recipes are special cases that only have code that describes how they work.

I had a thought during this PR (and my attempts to migrate some HTML content) that it'd be nice to specify, for example, expected frontmatter (whether required or optional). The recipe seems like a natural place for that, so I can imagine uses for recipe files even for kinds of recipes that don't get built the same way as ordinary reference pages.

Reply:
> Can we have recipe files that don't specify a body? Or even essentially emtpy recipe files that say, # read whatever.js to see how this recipe is handled?

It would be nice to have a single place to see what the valid recipes are, even if some recipes are special cases that only have code that describes how they work.

Yes, I think this is probably a better solution than a secret exception like we have at the moment. Part of the recipe thing, for me, is meta-documentation.

As I said, we could also perhaps make landing pages properly recipe-driven.

But this isn't a linter issue (just an issue that the linter uncovered :) so let's keep the linter behaviour as it is, to remind us to come back to it.